### ISSUE TITLE
[javascript] Possible issues in grammar of 'a=b=c=1+1'

### ISSUE THREAD

**Issue by yanmingsohu** [antlr/grammars-v4#1546]
bad:
a b =;
b c =;
(1+1) c =;

Right:
(1+1) c =;
b c =;
a b =;

**Comment by loonydev** [comment]
Hi @yanmingsohu, can u provide some example for testing?

**Comment by yanmingsohu** [comment]
> a=b=c=1+1

is real javascript code, in v8 engine, result is a=2, b=2, c=2
or you mean js engine c++ code?

**Comment by studentmain** [comment]
antlr: ((a=b)=c)=(1+1)
standard: a=(b=(c=(1+1)))

solution: add `<assoc=right>`

But I'd prefer a complete review for js grammar.

Jing Yanming <notifications@github.com> 于 2019年10月31日周四 下午8:38写道：

**Comment by yanmingsohu** [comment]
```diff
diff --git a/javascript/JavaScriptParser.g4 b/javascript/JavaScriptParser.g4
index 69b16cc..10660ab 100644
--- a/javascript/JavaScriptParser.g4
+++ b/javascript/JavaScriptParser.g4
@@ -351,8 +352,8 @@ singleExpression
     | singleExpression '&&' singleExpression                                 # LogicalAndExpression
     | singleExpression '||' singleExpression                                 # LogicalOrExpression
     | singleExpression '?' singleExpression ':' singleExpression             # TernaryExpression
-    | singleExpression '=' singleExpression                                  # AssignmentExpression
-    | singleExpression assignmentOperator singleExpression                   # AssignmentOperatorExpression
+    | <assoc=right> singleExpression '=' singleExpression                    # AssignmentExpression
+    | <assoc=right> singleExpression assignmentOperator singleExpression     # AssignmentOperatorExpression
     | singleExpression TemplateStringLiteral                                 # TemplateStringExpression  // ECMAScript 6
     | iteratorBlock                                                          # IteratorsExpression // ECMAScript 6
     | generatorBlock                                                         # GeneratorsExpression // ECMAScript 6
```

**Comment by KvanTTT** [comment]
Hi, thanks. Could you suggest a pull request with such a change?

**Comment by studentmain** [comment]
Can you check with other expression? They may have same problem.

BTW: This grammar is pretty old (maybe ES2016?). It's time updating it to ES2020 (working on it, no ETA, maybe in few days).
<hr/>

能不能检查一下其他表达式，估计也有这问题。

顺便这语法文件挺老了，估计是ES2016时代的，async/await都没有，是时候更新到ES2020了。

**Comment by KvanTTT** [comment]
I think your solution is good, because of right-associative assign expressions is common practice in all programming languages, including JavaScript.

> BTW: This grammar is pretty old (maybe ES2016?). It's time updating it to ES2020 (working on it, no ETA, maybe in few days).

Good news!

**Comment by KvanTTT** [comment]
fixed #1548
